File: benchmarks/scala/Benchmark.scala
import java.util.concurrent.{Executors, Callable, Future}

object Benchmark {

  // ---------------------------------------------------------------------------
  // Helpers
  // ---------------------------------------------------------------------------

  private def printResult(name: String, timeMs: Double, result: String): Unit =
    println(s"""{"language":"scala","benchmark":"$name","time_ms":$timeMs,"result":"$result"}""")

  private def timed[A](block: => A): (A, Double) = {
    val t0 = System.nanoTime()
    val res = block
    val t1 = System.nanoTime()
    (res, (t1 - t0) / 1e6)
  }

  // ---------------------------------------------------------------------------
  // 1. fib  --  naive recursive fibonacci(40)
  // ---------------------------------------------------------------------------

  private def fib(n: Int): Long =
    if (n < 2) n.toLong else fib(n - 1) + fib(n - 2)

  private def runFib(): Unit = {
    val (result, ms) = timed { fib(40) }
    printResult("fib", ms, result.toString)
  }

  // ---------------------------------------------------------------------------
  // 2. trees  --  binary tree depth 21, checksum
  // ---------------------------------------------------------------------------

  private case class TreeNode(left: Option[TreeNode], right: Option[TreeNode])

  private def buildTree(depth: Int): TreeNode =
    if (depth == 0) TreeNode(None, None)
    else TreeNode(Some(buildTree(depth - 1)), Some(buildTree(depth - 1)))

  private def checksum(node: TreeNode): Long = node match {
    case TreeNode(None, None) => 1L
    case TreeNode(Some(l), Some(r)) => 1L + checksum(l) + checksum(r)
    case _ => 1L
  }

  private def runTrees(): Unit = {
    val depth = 21
    val (result, ms) = timed {
      val tree = buildTree(depth)
      checksum(tree)
    }
    printResult("trees", ms, result.toString)
  }

  // ---------------------------------------------------------------------------
  // 3. matrix  --  1000x1000 Double matrix multiply
  // ---------------------------------------------------------------------------

  private def runMatrix(): Unit = {
    val n = 1000
    val a = Array.tabulate(n, n)((i, j) => (i + j).toDouble / n)
    val b = Array.tabulate(n, n)((i, j) => (i - j).toDouble / n)

    val (result, ms) = timed {
      val c = Array.ofDim[Double](n, n)
      var i = 0
      while (i < n) {
        var k = 0
        while (k < n) {
          val aik = a(i)(k)
          var j = 0
          while (j < n) {
            c(i)(j) += aik * b(k)(j)
            j += 1
          }
          k += 1
        }
        i += 1
      }
      c(0)(0)
    }
    printResult("matrix", ms, f"$result%.6f")
  }

  // ---------------------------------------------------------------------------
  // 4. strings  --  1 MB deterministic ASCII text processing
  // ---------------------------------------------------------------------------

  private def runStrings(): Unit = {
    // Build ~1 MB of deterministic ASCII text
    val words = Array("the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog",
      "alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta", "theta")
    val sb = new java.lang.StringBuilder(1100000)
    var idx = 0
    while (sb.length() < 1000000) {
      if (sb.length() > 0) sb.append(' ')
      sb.append(words(idx % words.length))
      idx += 1
    }
    val text = sb.toString

    val (wordCount, ms) = timed {
      // Count total words
      var wc = 0
      var inWord = false
      var i = 0
      val len = text.length
      while (i < len) {
        val c = text.charAt(i)
        if (c == ' ' || c == '\n' || c == '\t') {
          inWord = false
        } else {
          if (!inWord) { wc += 1; inWord = true }
        }
        i += 1
      }

      // Count occurrences of "the"
      var theCount = 0
      var pos = 0
      while (pos <= len - 3) {
        if (text.charAt(pos) == 't' && text.charAt(pos + 1) == 'h' && text.charAt(pos + 2) == 'e') {
          // make sure it's a whole word
          val before = pos == 0 || text.charAt(pos - 1) == ' '
          val after = pos + 3 >= len || text.charAt(pos + 3) == ' '
          if (before && after) theCount += 1
        }
        pos += 1
      }

      // Reverse the string
      val reversed = new java.lang.StringBuilder(text).reverse().toString

      // Simple hash (sum of chars)
      var hash = 0L
      i = 0
      while (i < reversed.length) {
        hash = hash * 31 + reversed.charAt(i)
        i += 1
      }

      wc
    }
    printResult("strings", ms, wordCount.toString)
  }

  // ---------------------------------------------------------------------------
  // 5. concurrent  --  ExecutorService, 1000 tasks computing fib(30), sum
  // ---------------------------------------------------------------------------

  private def runConcurrent(): Unit = {
    val numTasks = 1000
    val pool = Executors.newFixedThreadPool(Runtime.getRuntime.availableProcessors())

    val (result, ms) = timed {
      val futures = new Array[Future[Long]](numTasks)
      var i = 0
      while (i < numTasks) {
        futures(i) = pool.submit(new Callable[Long] {
          override def call(): Long = fib(30)
        })
        i += 1
      }

      var sum = 0L
      i = 0
      while (i < numTasks) {
        sum += futures(i).get()
        i += 1
      }
      sum
    }

    pool.shutdown()
    printResult("concurrent", ms, result.toString)
  }

  // ---------------------------------------------------------------------------
  // CLI dispatcher
  // ---------------------------------------------------------------------------

  private val benchmarks: Map[String, () => Unit] = Map(
    "fib"        -> (() => runFib()),
    "trees"      -> (() => runTrees()),
    "matrix"     -> (() => runMatrix()),
    "strings"    -> (() => runStrings()),
    "concurrent" -> (() => runConcurrent())
  )

  private val orderedNames = List("fib", "trees", "matrix", "strings", "concurrent")

  def main(args: Array[String]): Unit = {
    val targets = if (args.isEmpty || args.contains("all")) orderedNames
                  else args.toList.filter(benchmarks.contains)

    if (targets.isEmpty) {
      System.err.println("Usage: scala Benchmark <fib|trees|matrix|strings|concurrent|all>")
      sys.exit(1)
    }

    targets.foreach(name => benchmarks(name)())
  }
}

File: benchmarks/dart/benchmark.dart
import 'dart:isolate';
import 'dart:convert';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

void printResult(String name, double timeMs, String result) {
  final json = jsonEncode({
    'language': 'dart',
    'benchmark': name,
    'time_ms': double.parse(timeMs.toStringAsFixed(4)),
    'result': result,
  });
  print(json);
}

// ---------------------------------------------------------------------------
// 1. fib  --  naive recursive fibonacci(40)
// ---------------------------------------------------------------------------

int fib(int n) {
  if (n < 2) return n;
  return fib(n - 1) + fib(n - 2);
}

void benchFib() {
  final sw = Stopwatch()..start();
  final result = fib(40);
  sw.stop();
  printResult('fib', sw.elapsedMicroseconds / 1000.0, result.toString());
}

// ---------------------------------------------------------------------------
// 2. trees  --  binary tree depth 21, checksum
// ---------------------------------------------------------------------------

class TreeNode {
  final TreeNode? left;
  final TreeNode? right;

  TreeNode(this.left, this.right);

  int check() {
    if (left == null) return 1;
    return 1 + left!.check() + right!.check();
  }
}

TreeNode buildTree(int depth) {
  if (depth == 0) return TreeNode(null, null);
  return TreeNode(buildTree(depth - 1), buildTree(depth - 1));
}

void benchTrees() {
  const depth = 21;
  final sw = Stopwatch()..start();
  final tree = buildTree(depth);
  final checksum = tree.check();
  sw.stop();
  printResult('trees', sw.elapsedMicroseconds / 1000.0, checksum.toString());
}

// ---------------------------------------------------------------------------
// 3. matrix  --  1000x1000 double matrix multiply
// ---------------------------------------------------------------------------

void benchMatrix() {
  const n = 1000;

  // Allocate matrices as flat lists for cache friendliness.
  final a = List<double>.generate(n * n, (i) => (i ~/ n + 1) * 1.0);
  final b = List<double>.generate(n * n, (i) => (i % n + 1) * 1.0);
  final c = List<double>.filled(n * n, 0.0);

  final sw = Stopwatch()..start();
  for (int i = 0; i < n; i++) {
    final iOff = i * n;
    for (int k = 0; k < n; k++) {
      final aik = a[iOff + k];
      final kOff = k * n;
      for (int j = 0; j < n; j++) {
        c[iOff + j] += aik * b[kOff + j];
      }
    }
  }
  sw.stop();

  printResult('matrix', sw.elapsedMicroseconds / 1000.0, c[0].toString());
}

// ---------------------------------------------------------------------------
// 4. strings  --  1 MB deterministic ASCII text processing
// ---------------------------------------------------------------------------

void benchStrings() {
  // Build a deterministic ~1 MB ASCII string.
  const totalBytes = 1024 * 1024; // 1 MB
  final words = <String>[
    'the', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog',
    'alpha', 'beta', 'gamma', 'delta', 'epsilon', 'zeta', 'eta', 'theta',
  ];
  final buf = StringBuffer();
  int wi = 0;
  while (buf.length < totalBytes) {
    if (buf.length > 0) buf.write(' ');
    buf.write(words[wi % words.length]);
    wi++;
  }
  final text = buf.toString();

  final sw = Stopwatch()..start();

  // Count words
  int wordCount = 0;
  bool inWord = false;
  for (int i = 0; i < text.length; i++) {
    final c = text.codeUnitAt(i);
    if (c == 0x20 || c == 0x0A || c == 0x0D || c == 0x09) {
      inWord = false;
    } else {
      if (!inWord) wordCount++;
      inWord = true;
    }
  }

  // Count occurrences of "the"
  int theCount = 0;
  int idx = 0;
  while (true) {
    idx = text.indexOf('the', idx);
    if (idx == -1) break;
    // Check word boundary (space or start/end)
    final before = idx == 0 || text.codeUnitAt(idx - 1) == 0x20;
    final after = idx + 3 >= text.length || text.codeUnitAt(idx + 3) == 0x20;
    if (before && after) theCount++;
    idx += 3;
  }

  // Reverse the string
  final reversed = String.fromCharCodes(text.runes.toList().reversed);

  // Simple hash (djb2)
  int hash = 5381;
  for (int i = 0; i < reversed.length; i++) {
    hash = ((hash << 5) + hash + reversed.codeUnitAt(i)) & 0x7FFFFFFF;
  }

  sw.stop();

  // Use theCount and hash to prevent dead-code elimination (via identity).
  _ = theCount;
  _ = hash;

  printResult('strings', sw.elapsedMicroseconds / 1000.0, wordCount.toString());
}

// Prevent tree-shaking / dead-code elimination.
dynamic _ = null;

// ---------------------------------------------------------------------------
// 5. concurrent  --  isolates computing fib(30)
//    100 isolates, each computing fib(30) x 10 = 1000 total fibs
// ---------------------------------------------------------------------------

int fibIsolateWork(int unused) {
  int total = 0;
  for (int i = 0; i < 10; i++) {
    total += fib(30);
  }
  return total;
}

Future<void> benchConcurrent() async {
  const numIsolates = 100;

  final sw = Stopwatch()..start();

  final futures = <Future<int>>[];
  for (int i = 0; i < numIsolates; i++) {
    futures.add(Isolate.run(() => fibIsolateWork(i)));
  }
  final results = await Future.wait(futures);
  final sum = results.fold<int>(0, (a, b) => a + b);

  sw.stop();
  printResult('concurrent', sw.elapsedMicroseconds / 1000.0, sum.toString());
}

// ---------------------------------------------------------------------------
// CLI dispatcher
// ---------------------------------------------------------------------------

Future<void> main(List<String> args) async {
  final benchmarks = <String, Future<void> Function()>{
    'fib': () async => benchFib(),
    'trees': () async => benchTrees(),
    'matrix': () async => benchMatrix(),
    'strings': () async => benchStrings(),
    'concurrent': () async => await benchConcurrent(),
  };

  if (args.isEmpty) {
    print('Usage: benchmark <name|all>');
    print('Available: ${benchmarks.keys.join(", ")}, all');
    return;
  }

  final requested = args[0];

  if (requested == 'all') {
    for (final entry in benchmarks.entries) {
      await entry.value();
    }
  } else if (benchmarks.containsKey(requested)) {
    await benchmarks[requested]!();
  } else {
    print('Unknown benchmark: $requested');
    print('Available: ${benchmarks.keys.join(", ")}, all');
  }
}
